### PR TITLE
Allow string keys in `eval` utility

### LINKED ISSUE #227: Allow string keys in `.eval`
### Description

`eval` is very useful for debugging, including interactive debugging during development. One issue is that required inputs must be passed by reference which is not always easy to do inside a given stack. 

I usually have to play with ancestors to get a hold of the input variable. Instead it would be neat if we can get it by name if a string is used as a key in `eval`

```python
import pytensor.tensor as pt

def create_expr():
  x = pt.scalar("x")
  y = pt.exp(x)
  return y

y = create_expr()
# Feature request:
y.eval({"x": 0})  # 1.0
```

If there are multiple variables with the same key name we could use the closest + warning or raise.

There is already a handy utility to find variables by name: https://pytensor.readthedocs.io/en/latest/library/graph/graph.html#pytensor.graph.basic.get_var_by_name

### PR DESCRIPTION
**Pull request by Raj-Parekh24**
### Motivation for these changes
Closes #227 

### Implementation details
 I have implemented a function that converts the dictionary with string keys to pytensor. The pytensor variable for the corresponding name is found using the function "get_var_by_name".

### Checklist
+ [x] Explain motivation and implementation 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues, preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes). Note that if they don't, we will [rewrite/rebase/squash the git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) before merging.
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## New features
- ... Its and enhancement of current functionality eval present in graph/basic.py under Variable class. Through this enhancement eval will accept string i.e. pytensor node name. 

## Bugfixes
- ...

## Documentation
- ...

## Maintenance
- ...

### CODE REVIEW

File: tests/graph/test_basic.py [hunk 1]
```diff
@@ -302,6 +302,14 @@ def test_eval(self):
...
         ), "temporary functions must not be serialized"
 
+    def test_eval_with_strings(self):
+        assert self.w.eval({"x": 1.0, "y": 2.0}) == 6.0
+        assert self.w.eval({self.z: 3}) == 6.0
+        assert hasattr(self.w, "_fn_cache"), "variable must have cache after eval"
+        assert not hasattr(
+            pickle.loads(pickle.dumps(self.w)), "_fn_cache"
+        ), "temporary functions must not be serialized"
+
```
Comment (ricardoV94) [inline tests/graph/test_basic.py:312]: I don't think we need these checks. Instead more useful would be to include a mixed case like `.eval({"x": 1.0, y: 2.0})` and check the cases where the string does not match any variable name.

Finally we may want to add one test case with repeated variable names such as:
```python
x = pt.scalar()
x.name="x"
x = x + 1
x.name = "x"
y = pt.exp(x)
y.eval({"x": 0})
```
I think we should fail in those cases. Does `get_var_by_name` return all the instances with the name?

**Review by ricardoV94 — COMMENTED** [review]
This looks great @Raj-Parekh24 

I left some suggestions below

File: tests/graph/test_basic.py (hunk 1 above)
Comment (Raj-Parekh24) [reply tests/graph/test_basic.py:312]: Hi @ricardoV94 ,

Thank you for the review.

`get_var_by_name` returns all instances of the same name. I am choosing the first declared variable i.e. last variable from the list to replace it with the name key.

**Comment by Raj-Parekh24** [comment]
While running `pre-commit run --all` in the local machine I am getting a series of errors with the last message as `142/177 files pass as expected.`. But when I check its logs for basic.py there are no errors from my changes.

Is this expected behavior or I have broken something?

**Comment by Raj-Parekh24** [comment]
And in the last build, the test case failed for multiple environments is:
`FAILED tests/link/jax/test_scalar.py::test_erfinv - AssertionError: `

**Comment by ricardoV94** [comment]
> And in the last build, the test case failed for multiple environments is:
> `FAILED tests/link/jax/test_scalar.py::test_erfinv - AssertionError: `

Yeah needs to be investigated. Probably a new release of jax that behaves differently.

The tracking issue is https://github.com/pymc-devs/pytensor/issues/241

**Comment by ricardoV94** [comment]
Mypy doesn't work on python 3.11, issue here https://github.com/pymc-devs/pytensor/issues/240

**Comment by Raj-Parekh24** [comment]
Thanks for the help @ricardoV94 ,

I have updated the code and added warning when multiple variables are there for same name.

And also updated the test cases accordingly.

Please review it.

File: pytensor/graph/basic.py [hunk 2]
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
             return [self.owner]
         return []
 
+    def convert_string_keys_to_pytensor_variables(self, inputs_to_values):
```
Comment (ricardoV94) [inline pytensor/graph/basic.py:561]: Let's make this an internal function of `eval`, no need to have it accessible to other functions in an outer scope

File: pytensor/graph/basic.py [hunk 3]
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
...
+            if isinstance(i, str):
+                nodes_with_matching_names = get_var_by_name([self], i)
+                length_of_nodes_with_matching_names = len(nodes_with_matching_names)
+                if length_of_nodes_with_matching_names == 0:
+                    raise Exception(f"{i} not found in graph")
+                else:
+                    if length_of_nodes_with_matching_names > 1:
+                        warnings.warn(
+                            f"Found {length_of_nodes_with_matching_names} pytensor variables with name {i} taking the first declared named variable for computation"
+                        )
```
Comment (ricardoV94) [inline pytensor/graph/basic.py:591]: It seems safer to just fail.

File: pytensor/graph/basic.py [hunk 4]
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
...
+        Examples
+        --------
+
+        >>> import numpy as np
+        >>> import pytensor.tensor as at
+        >>> x = at.dscalar('x')
+        >>> y = at.dscalar('y')
+        >>> z = x + y
+        >>> np.allclose(z.eval({'x' : 3, 'y' : 1}), 4)
+        True
```
Comment (ricardoV94) [inline pytensor/graph/basic.py:578]: This does not show a use of the function. In any case we don't need a docstring with an example I think

File: pytensor/graph/basic.py (hunk 2 above)
Comment (ricardoV94) [inline pytensor/graph/basic.py:561]: Given we are in `pytensor`, no need to mention it in the name. If you want you can specify the return type is `Dict[Variable, Variable]`

```suggestion
    def convert_string_keys_to_variables(self, inputs_to_values):
```

**Comment by ricardoV94** [comment]
@Raj-Parekh24 nice work. I left some comments above. Let me know what you think.

File: pytensor/graph/basic.py (hunk 2 above)
Comment (Raj-Parekh24) [reply pytensor/graph/basic.py:561]: Ok then I will insert logic within the implementation part only.

File: pytensor/graph/basic.py (hunk 3 above)
Comment (Raj-Parekh24) [reply pytensor/graph/basic.py:591]: Instead of warning will throw Exception instead.

File: pytensor/graph/basic.py (hunk 4 above)
Comment (Raj-Parekh24) [reply pytensor/graph/basic.py:578]: Since we are making it internal to eval I also think that the doc-string will be no more required.

File: tests/graph/test_basic.py [hunk 5]
```diff
@@ -302,6 +304,13 @@ def test_eval(self):
             pickle.loads(pickle.dumps(self.w)), "_fn_cache"
         ), "temporary functions must not be serialized"
 
+    def test_eval_with_strings(self):
+        assert self.w.eval({"x": 1.0, self.y: 2.0}) == 6.0
+        assert self.w.eval({self.z: 3}) == 6.0
+
+    def test_eval_with_strings_with_mulitple_same_name(self):
+        assert self.t.eval({"e": 1.0}) == 2.0
```
[the remark this answers is not in the document]
Comment (Raj-Parekh24) [reply tests/graph/test_basic.py:312]: ok, instead of creating variables in setup_method will add the variables in test_function.

**Comment by Raj-Parekh24** [comment]
Hi @ricardoV94 ,

Thanks for the review and the suggestions, I have updated the implementation accordingly and pushed a new commit regarding it. Please review it and suggest the required changes.

**Comment by Raj-Parekh24** [comment]
Hey @ricardoV94 ,

I have updated the code bases on your suggestions. 

And I also want to work on issue #55, I got the requirements but still not able to understand how to implement it. Since I am new to the repo, can you please give me starting steps to work on it.

**Comment by ricardoV94** [comment]
Hi @Raj-Parekh24 sorry for the delay. I'll try to review your PR tomorrow. For that other PR I will give some more pointers there

Thanks for all the work!

**Comment by Raj-Parekh24** [comment]
Thanks for all the help and support @ricardoV94 .

File: tests/graph/test_basic.py [hunk 6]
```diff
@@ -302,6 +302,17 @@ def test_eval(self):
             pickle.loads(pickle.dumps(self.w)), "_fn_cache"
         ), "temporary functions must not be serialized"
 
+    def test_eval_with_strings(self):
+        assert self.w.eval({"x": 1.0, self.y: 2.0}) == 6.0
+        assert self.w.eval({self.z: 3}) == 6.0
+
+    def test_eval_errors_having_mulitple_variables_same_name(self):
```
Comment (ricardoV94) [inline tests/graph/test_basic.py:309]: Should also add a test with zero variables matching since you have a specific error message for that case.

**Review by ricardoV94 — COMMENTED** [review]
Looks good. I left some suggestions below

File: tests/graph/test_basic.py (hunk 6 above)
Comment (Raj-Parekh24) [reply tests/graph/test_basic.py:309]: Thank you for your review @ricardoV94. Will add the corresponding test case.

File: pytensor/graph/basic.py [hunk 7]
```diff
@@ -597,6 +597,30 @@ def eval(self, inputs_to_values=None):
...
+                        raise Exception(f"{i} not found in graph")
+                    else:
+                        if length_of_nodes_with_matching_names > 1:
+                            raise Exception(
+                                f"Found {length_of_nodes_with_matching_names} pytensor variables with name {i}"
+                            )
+                        process_input_to_values[
+                            nodes_with_matching_names[
+                                length_of_nodes_with_matching_names - 1
+                            ]
```
[the remark this answers is not in the document]
Comment (Raj-Parekh24) [reply pytensor/graph/basic.py]: Thank you for your review @ricardoV94. Will update the index value.

**Comment by Raj-Parekh24** [comment]
Thanks a lot for your review @ricardoV94,

I have added the code based on what you suggested. Please review it and kindly give your suggestions.

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Functionality-wise this PR is ready! I just left some comments on code formatting. Let me know if you have any questions.

**Comment by Raj-Parekh24** [comment]
Thanks for the review @ricardoV94 ,

I have incorporated the suggested changes in the new commit, please review it.

**Review by ricardoV94 — APPROVED** [review]
Looks good, running the tests now!
